CRM/Mailing/ActionTokens.php

```diff
@@ -0,0 +1,121 @@
+<?php
+
+/*
+ +--------------------------------------------------------------------+
+ | CiviCRM version 4.7                                                |
+ +--------------------------------------------------------------------+
+ | Copyright CiviCRM LLC (c) 2004-2016                                |
+ +--------------------------------------------------------------------+
+ | This file is a part of CiviCRM.                                    |
+ |                                                                    |
+ | CiviCRM is free software; you can copy, modify, and distribute it  |
+ | under the terms of the GNU Affero General Public License           |
+ | Version 3, 19 November 2007 and the CiviCRM Licensing Exception.   |
+ |                                                                    |
+ | CiviCRM is distributed in the hope that it will be useful, but     |
+ | WITHOUT ANY WARRANTY; without even the implied warranty of         |
+ | MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.               |
+ | See the GNU Affero General Public License for more details.        |
+ |                                                                    |
+ | You should have received a copy of the GNU Affero General Public   |
+ | License and the CiviCRM Licensing Exception along                  |
+ | with this program; if not, contact CiviCRM LLC                     |
+ | at info[AT]civicrm[DOT]org. If you have questions about the        |
+ | GNU Affero General Public License or the licensing of CiviCRM,     |
+ | see the CiviCRM license FAQ at http://civicrm.org/licensing        |
+ +--------------------------------------------------------------------+
+ */
+
+/**
+ * Class CRM_Mailing_ActionTokens
+ *
+ * Generate "action.*" tokens for mailings.
+ *
+ * To activate these tokens, the TokenProcessor context must specify:
+ * "mailingJobId" (int)
+ * "mailingActionTarget" (array) with keys:
+ *   'id' => int, event queue ID
+ *   'hash' => string, event queue hash code
+ *   'contact_id' => int, contact_id,
+ *   'email' => string, email
+ *   'phone' => string, phone
+ */
+class CRM_Mailing_ActionTokens extends \Civi\Token\AbstractTokenSubscriber {
+
+  /**
+   * Class constructor.
+   */
+  public function __construct() {
+    // TODO: Think about supporting dynamic tokens like "{action.subscribe.\d+}"
+    parent::__construct('action', array(
+      'subscribeUrl' => ts('Subscribe URL (Action)'),
+      'forward' => ts('Forward URL (Action)'),
+      'optOut' => ts('Opt-Out (Action)'),
+      'optOutUrl' => ts('Opt-Out URL (Action)'),
+      'reply' => ts('Reply (Action)'),
+      'unsubscribe' => ts('Unsubscribe (Action)'),
+      'unsubscribeUrl' => ts('Unsubscribe URL (Action)'),
+      'resubscribe' => ts('Resubscribe (Action)'),
+      'resubscribeUrl' => ts('Resubscribe URL (Action)'),
+      'eventQueueId' => ts('Event Queue ID'),
+    ));
+  }
+
+  /**
+   * Evaluate the content of a single token.
```

Comment:
Hmm, well, I have pretty mixed feelings about this, e.g.

 * The tool I use most (PHPStorm) doesn't do anything special with `inheritDoc`, so it's easier to read the docs if you copy the docblock. (Would be *really* nice to cmd-click or hover or something to see the docs in situ.)
 * Copying sucks. (Duplication/stale info/etal.)
 * Civi coding standards (de jure) extend Drupal standards as a baseline, which means we ought to follow https://www.drupal.org/docs/develop/coding-standards/api-documentation-and-comment-standards#inheritdoc -- i.e. `{@inheritDoc}`
 * Civi code standards (de facto) use `@inheritDoc` instead of `{@inheritDoc}`.
 * The vast majority of cases of `@inheritDoc` are redundant/useless (you can just omit the docblock and [the docs say it will assume inheritance](https://www.phpdoc.org/docs/latest/guides/inheritance.html)).

Do you see a concrete upside in resolving this? What is it?